flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPoolTest.java

```diff
@@ -518,4 +528,176 @@ public void go() throws Exception {
+				@Override
+				public void go() throws Exception {
+					exclusiveSegments.addAll(globalPool.requestMemorySegments());
+				}
+			};
+			asyncRequest.start();
+
+			// wait until no buffer is available
+			final Deadline deadline = Deadline.fromNow(Duration.ofSeconds(10L));
+			while (globalPool.getNumberOfAvailableMemorySegments() > 0) {
+				Thread.sleep(50);
+				if (!deadline.hasTimeLeft()) {
+					fail("Waiting timeout.");
+				}
+			}
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(numberOfSegmentsToRequest, exclusiveSegments.size());
+
+			for (int i = 1; i < numberOfSegmentsToRequest; ++i) {
+				globalPool.recycle(segments.get(i));
+			}
+			segments.clear();
+
+			asyncRequest.sync();
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(numBuffers, exclusiveSegments.size());
+			assertFalse(globalPool.isAvailable().isDone());
+
+			globalPool.recycleMemorySegments(exclusiveSegments);
+			exclusiveSegments.clear();
+			assertTrue(globalPool.isAvailable().isDone());
+			assertEquals(numBuffers, globalPool.getNumberOfAvailableMemorySegments());
+
+		} finally {
+			globalPool.destroy();
+		}
+	}
+
+	/**
+	 * Tests that blocking request of multi local buffer pools can be fulfilled by recycled segments
+	 * to the global network buffer pool.
+	 */
+	@Test
+	public void testBlockingRequestFromMultiLocalBufferPool() throws Exception {
+		final int localBufferPoolSize = 10;
+		final int numLocalBufferPool = 10;
+		final int numBuffers = numLocalBufferPool * localBufferPoolSize;
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(numLocalBufferPool);
+		final NetworkBufferPool globalPool = new NetworkBufferPool(numBuffers, 128, 5);
+
+		try {
+			// create local buffer pools
+			final List<BufferPool> localBufferPools = new ArrayList<>(numLocalBufferPool);
+			for (int i  = 0; i < numLocalBufferPool; ++i) {
+				localBufferPools.add(
+					globalPool.createBufferPool(localBufferPoolSize, Integer.MAX_VALUE));
+			}
+
+			// request some buffers from the global pool
+			final List<MemorySegment> segments = new ArrayList<>(numBuffers / 2);
+			for (int i = 0; i < numBuffers / 2; ++i) {
+				segments.add(globalPool.requestMemorySegment());
+			}
+			assertEquals(numBuffers - segments.size(), globalPool.getNumberOfAvailableMemorySegments());
+			assertTrue(globalPool.isAvailable().isDone());
+
+			// blocking request buffers form local buffer pools
+			final CountDownLatch latch = new CountDownLatch(numLocalBufferPool);
+			final AtomicReference<Throwable> cause = new AtomicReference<>();
+			final BlockingQueue<BufferBuilder> segmentsRequested = new ArrayBlockingQueue<>(numBuffers);
+			for (final BufferPool bufferPool: localBufferPools) {
+				executorService.submit(() -> {
+					for (int num = localBufferPoolSize; num > 0; --num) {
+						try {
+							segmentsRequested.add(bufferPool.requestBufferBuilderBlocking());
+						} catch (Exception e) {
+							cause.set(e);
```

Comment:
The `cause` is actually not used after setting, and we should throw this exception.